syntax = "proto3";

import "google/protobuf/timestamp.proto";

import "bob_emploi/frontend/api/user.proto";

package bayes.bob;

message UseCasePools {
  // A sorted list of use case groups. The ones available in the db to be fetch.
  repeated UseCasePool use_case_pools = 1;
}

message UseCases {
  // A list of anonymized user examples.
  repeated UseCase use_cases = 1;
}

message UseCasePool {
  // Name of the pool, usually the date of creation.
  string name = 1;

  // How many use cases are in this group.
  int32 useCaseCount = 2;

  // How many use cases have already been evaluated in this group.
  int32 evaluatedUseCaseCount = 3;

  // Timestamp of the most recent use case.
  google.protobuf.Timestamp last_user_registered_at = 4;
}

message UseCase {
  // Unique use case ID.
  string use_case_id = 1;

  // Name of the use case pool it is part of, if any.
  string pool_name = 2;

  // Index of this use case in its pool.
  int32 index_in_pool = 3;

  // Title of the use case, e.g. "Baker in Rennes";
  string title = 4;

  // Actual user data.
  User user_data = 5;

  // The evaluation for this use case.
  UseCaseEvaluation evaluation = 6;
}

message UseCaseEvaluation {
  // When the evaluation was last done.
  google.protobuf.Timestamp evaluated_at = 1;

  // Email address of the user that did the latest evaluation of the use case.
  string by = 6;

  // General score for this use case.
  UseCaseScore score = 2;

  // General comments for this use case.
  string comments = 3;

  // New score for advice modules.
  // TODO: put this info in AdviceEvaluation if possible (note that we need to store "0" value
  // for "modules" to mean that the advice should be hidden).
  map<string, int32> modules = 4;

  // Evaluation for each advice modules.
  map<string, AdviceEvaluation> advices = 5;
}

message AdviceEvaluation {
  // Whether the advice needs to be optimized with better data and content.
  bool shouldBeOptimized = 1;

  // Comment about the advice.
  string comment = 2;
}

enum UseCaseScore {
  UNKOWN_EVALUATION = 0;
  EXCELLENT = 1;
  GOOD_ENOUGH = 2;
  BAD = 3;
}

message UseCaseCreateRequest {
  // Email of the user to create the use case from.
  string email = 1;

  // Name of the pool in which to add the use case.
  string pool_name = 2;
}
